### PR TITLE
baggage: Member.String encodes only when necessary

### PR DESCRIPTION
**Pull request by pellared**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4773

I also added [Add UTF-8 test cases](https://github.com/open-telemetry/opentelemetry-go/pull/4775/commits/c117cd7a06d946c1cae48bd277d0d42065dfb55d) and some benchmarks.

Benchmark:

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/baggage
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
BenchmarkNew-16                   668466              1628 ns/op             832 B/op         16 allocs/op
BenchmarkNewMember-16           26077026                45.27 ns/op            0 B/op          0 allocs/op
BenchmarkParse-16                 954313              1116 ns/op             832 B/op          7 allocs/op
BenchmarkString-16                569619              1893 ns/op             960 B/op         18 allocs/op
BenchmarkValueEscape/nothing_to_escape-16               135047982                7.684 ns/op           0 B/op          0 allocs/op
BenchmarkValueEscape/requires_escaping-16               28184155                42.72 ns/op           16 B/op          1 allocs/op
BenchmarkValueEscape/long_value-16                       1492291               987.6 ns/op           576 B/op          2 allocs/op
```

Benchstat (in `old.txt` the benchmarks use `url.PathEscape` instead of `valueEscape`):

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/baggage
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
                                 │   old.txt    │               new.txt               │
                                 │    sec/op    │   sec/op     vs base                │
New-16                              1.221µ ± 5%   1.262µ ± 7%        ~ (p=0.055 n=10)
NewMember-16                        35.38n ± 5%   35.50n ± 3%        ~ (p=0.811 n=10)
Parse-16                            963.8n ± 4%   929.9n ± 3%   -3.52% (p=0.019 n=10)
String-16                           2.183µ ± 6%   1.632µ ± 4%  -25.22% (p=0.000 n=10)
ValueEscape/nothing_to_escape-16   13.465n ± 6%   6.957n ± 4%  -48.33% (p=0.000 n=10)
ValueEscape/requires_escaping-16    67.95n ± 5%   36.34n ± 6%  -46.52% (p=0.000 n=10)
ValueEscape/long_value-16          1596.0n ± 5%   739.2n ± 5%  -53.68% (p=0.000 n=10)
geomean                             279.3n        199.8n       -28.48%

                                 │    old.txt    │               new.txt                │
                                 │     B/op      │    B/op     vs base                  │
New-16                              832.0 ± 0%     832.0 ± 0%        ~ (p=1.000 n=10) ¹
NewMember-16                        0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
Parse-16                            832.0 ± 0%     832.0 ± 0%        ~ (p=1.000 n=10) ¹
String-16                          1024.0 ± 0%     960.0 ± 0%   -6.25% (p=0.000 n=10)
ValueEscape/nothing_to_escape-16    0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
ValueEscape/requires_escaping-16    16.00 ± 0%     16.00 ± 0%        ~ (p=1.000 n=10) ¹
ValueEscape/long_value-16           640.0 ± 0%     576.0 ± 0%  -10.00% (p=0.000 n=10)
geomean                                        ²                -2.40%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                 │   old.txt    │               new.txt               │
                                 │  allocs/op   │ allocs/op   vs base                 │
New-16                             16.00 ± 0%     16.00 ± 0%       ~ (p=1.000 n=10) ¹
NewMember-16                       0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Parse-16                           7.000 ± 0%     7.000 ± 0%       ~ (p=1.000 n=10) ¹
String-16                          18.00 ± 0%     18.00 ± 0%       ~ (p=1.000 n=10) ¹
ValueEscape/nothing_to_escape-16   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
ValueEscape/requires_escaping-16   1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
ValueEscape/long_value-16          2.000 ± 0%     2.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                       ²               +0.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

The results of `String` and `ValueEscape` show improvement.

### CODE REVIEW

**Comment by dmathieu** [comment]
Should you run benchstat to compare the benchmarks between main and your branch?

File: baggage/baggage.go [hunk 1]
```diff
@@ -652,9 +652,65 @@ func validateValue(s string) bool {
...
-		c == 0x21
+		(c >= 0x5d && c <= 0x7e)
+}
+
+// valueEscape escapes the string so it can be safely placed inside a baggage value,
+// replacing special characters with %XX sequences as needed.
+//
+// The implementation is based on:
+// https://github.com/golang/go/blob/f6509cf5cdbb5787061b784973782933c47f1782/src/net/url/url.go#L285.
+func valueEscape(s string) string {
```
Comment (pellared) [inline baggage/baggage.go:667]: I think we can export this function in order to have better resolution for https://github.com/open-telemetry/opentelemetry-go/issues/3685.

Of course the user would be able to call 

```go
baggage.NewMember(key, baggage.ValueEscape(val))
```

The benefit of it would be that the change would be backwards compatible.

I prefer to do this in a separate PR as I think it would be good to also update the documentation.

Comment (pellared) [reply baggage/baggage.go:667]: FYI @yurishkuro

**Comment by pellared** [comment]
> Should you run benchstat to compare the benchmarks between main and your branch?

I can do that today in a few hours or tomorrow morning. 
The result should be similar (maybe a little faster as there should be a few CPU cycles less). I intentionally almost copied the code from `url.PathEscape`. I wanted to add the benchmarks if someone would want to improve it in future.

File: baggage/baggage.go (hunk 1 above)
Comment (yurishkuro) [reply baggage/baggage.go:667]: No objections to this change, but I don't see how it would fix that other issue.

Comment (pellared) [reply baggage/baggage.go:667]: It would not be a direct fix, but I think it would be a step towards resolving it.

E.g. I think that `baggage.ValueEscape` would have to be used here:

https://github.com/open-telemetry/opentelemetry-go/blob/885210bf33238f8c6cc94c3f3711b0036fbe77b1/bridge/opentracing/bridge.go#L78

Comment (yurishkuro) [reply baggage/baggage.go:667]: The main issue in #3685 is not about OT Bridge, but about a fundamental misunderstanding of the difference between Baggage API and W3C Baggage encoding rules. IMO it's an outright bug in this implementation. The Spec, unfortunately, was not prescriptive enough about that.

**Comment by pellared** [comment]
@dmathieu Benchstat results added.

**Review by dmathieu — APPROVED** [review]

**Review by yurishkuro — APPROVED** [review]
I have no objection to this as an optimization of _existing incorrect behavior_ - its not supposed to encode anything at all, only Propagator should be doing the encoding.

**Comment by MrAlias** [comment]
This was merged without two approvals from the Go Approvals team.
